pkg/registry/core/service/strategy.go

```diff
@@ -226,3 +218,19 @@ func (serviceStatusStrategy) PrepareForUpdate(ctx context.Context, obj, old runt
 func (serviceStatusStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Object) field.ErrorList {
 	return validation.ValidateServiceStatusUpdate(obj.(*api.Service), old.(*api.Service))
 }
+
+func clearClusterIPRelatedFields(newService, oldService *api.Service) {
+	// in order to allow existing updates to service (specifically converting service from ClusterIP => ExternalName) to work as is
+	// we clear the new optional (fields IPFamilies and IPFamily) Policy on behalf of the user
+	if newService.Spec.Type == api.ServiceTypeExternalName && oldService.Spec.Type != api.ServiceTypeExternalName {
+		// if ClusterIP field was clear, then we clear ClusterIPs as well
+		// _that means_ for dual stack services (i.e. len(ClusterIPs) > 1), user have to clear ClusterIPs
+		// since we operate on core (not v1), we have detect this via ClusterIPs[0]
+		if len(newService.Spec.ClusterIPs) == 0 || (len(newService.Spec.ClusterIPs[0]) == 0 && len(newService.Spec.ClusterIPs) == 1) {
```

Comment:
Are you sure this is right?  Conversion only copies clusterIP -> clusterIPs[0] if len() > 0. as far as I can tell.

Reply:
hmm this is problematic, when converting to ExternalName
##for clients who are not dualstack aware
```
v1 
ClusterIP:"" // cleared by user
ClusterIPs: "(old-ip)" // Because it will not copy it len() == 0

resulting into core that looks like
core
ClusterIPs: {"old-ip"}
```

##for clients that are using only ClusterIPs
```
v1
ClusterIP:"old-IP"
ClusterIPs: nil // cleared by user

resultant core:
ClusterIPs: "old-ip" // because it is authoritative
```

> the above problem also occurs with `headless` conversion.

Reply:
One way to solve this is to assume that even new clients will have to set ClusterIP and ClusterIPs *and* we consider ClusterIP authoritative even if it is empty strings. This will be the case until we drop ClusterIP in some distant (super distant) future version.

Reply:
For old clients, you have code for this (unless I misunderstand):

```
func Convert_v1_ServiceSpec_To_core_ServiceSpec(...) {
    //...

    // when user converts to ExternalName they set .spec.clusterIP = ""
    // but current scripts will not set .spec.clusterIPs to nil
    // the below fixes this
    if out.Type == core.ServiceTypeExternalName {
        if len(in.ClusterIP) == 0 && len(out.ClusterIPs) > 0 { 
            // reset clusterIPs
            out.ClusterIPs = nil 
        }
    }   
    
    //...
}
```

Now, I will be the first to say that this is suspect.  We don't generally mutate user input and we can't (here) tell if the user is trying to CREATE something with both ExternalName && ClusterIPs[] (should fail) or trying to update to ExternalName and is old.

It seems reasonable to clear it here, but it does set a precedent and *could* result in confusion.  E.g. a bad YAML might be continually re-applied because server has empty ClusterIPs, but YAML has non-empty.

For new clients, it's less obvious that we can infer the intention.  Specifically you also have code in conversion:

```
     if len(in.ClusterIP) > 0 && len(in.ClusterIPs) == 0 {
        out.ClusterIPs = []string{in.ClusterIP}
    }
```

So by the time we get into registry, we will have "fixed" the problem in the wrong direction.  They have to specify both fields.

The alternative here would be to consider old and new values when making these mutations, which I am increasingly thinking we should do.  I just don't know the right hook for that.

Reply:
> The alternative here would be to consider old and new values when making these mutations, which I am increasingly thinking we should do. I just don't know the right hook for that.

Is there some way we can make use of `conversion.Scope`?

Reply:
I talked to @lavalamp and he suggested PrepareForUpdate() as an alternative.  I starte to mock it out in https://github.com/kubernetes/kubernetes/compare/master...thockin:dual-plurals-thockin-hack?expand=1 which takes @khenidak first 2 commits and then adds my own on top.  Not at all tested yet.